pgwire: v23.1.8: error in argument for ×: unsupported OID 100446 with format code FormatText

This issue was auto filed by Sentry. It represents a crash or reported error on a live cluster with telemetry enabled.

Sentry Link: [https://cockroach-labs.sentry.io/issues/4597232051/?referrer=webhooks_plugin](https://cockroach-labs.sentry.io/issues/4597232051/?referrer=webhooks_plugin)

Panic Message:

```
encoding.go:842: error in argument for ×: unsupported OID 100446 with format code FormatText
(1) candidate pg code: 08P01
Wraps: (2) attached stack trace
  -- stack trace:
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execBind.func2
  | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_prepare.go:469
  | [...repeated from below...]
Wraps: (3) error in argument for ×
Wraps: (4) assertion failure
Wraps: (5) attached stack trace
  -- stack trace:
  | github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgwirebase.DecodeDatum
  | 	github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgwirebase/encoding.go:842
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execBind.func2
  | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_prepare.go:461
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execBind
  | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_prepare.go:478
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd
  | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2319
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run
  | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2103
  | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn
  | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:900
  | github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1
  | 	github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:310
  | runtime.goexit
  | 	src/runtime/asm_arm64.s:1172
Wraps: (6) unsupported OID 100446 with format code FormatText
Error types: (1) *pgerror.withCandidateCode (2) *withstack.withStack (3) *errutil.withPrefix (4) *assert.withAssertionFailure (5) *withstack.withStack (6) *errutil.leafError
-- report composition:
*errutil.leafError: unsupported OID 100446 with format code FormatText
encoding.go:842: *withstack.withStack (top exception)
*assert.withAssertionFailure
*errutil.withPrefix: error in argument for ×
conn_executor_prepare.go:469: *withstack.withStack (1)
*pgerror.withCandidateCode: 08P01
(check the extra data payloads)
```

<details>
<summary>Stacktrace (expand for inline code snippets):</summary>

[src/runtime/asm_arm64.s#L1171-L1173](src/runtime/asm_arm64.s#L1171-L1173)
[https://github.com/cockroachdb/cockroach/blob/2c1d4afd2c2de091a00a4dba4392ceeaf8504a95/pkg/sql/pgwire/conn.go#L309-L311](https://github.com/cockroachdb/cockroach/blob/2c1d4afd2c2de091a00a4dba4392ceeaf8504a95/pkg/sql/pgwire/conn.go#L309-L311)
[https://github.com/cockroachdb/cockroach/blob/2c1d4afd2c2de091a00a4dba4392ceeaf8504a95/pkg/sql/conn_executor.go#L899-L901](https://github.com/cockroachdb/cockroach/blob/2c1d4afd2c2de091a00a4dba4392ceeaf8504a95/pkg/sql/conn_executor.go#L899-L901)
[https://github.com/cockroachdb/cockroach/blob/2c1d4afd2c2de091a00a4dba4392ceeaf8504a95/pkg/sql/conn_executor.go#L2102-L2104](https://github.com/cockroachdb/cockroach/blob/2c1d4afd2c2de091a00a4dba4392ceeaf8504a95/pkg/sql/conn_executor.go#L2102-L2104)
[https://github.com/cockroachdb/cockroach/blob/2c1d4afd2c2de091a00a4dba4392ceeaf8504a95/pkg/sql/conn_executor.go#L2318-L2320](https://github.com/cockroachdb/cockroach/blob/2c1d4afd2c2de091a00a4dba4392ceeaf8504a95/pkg/sql/conn_executor.go#L2318-L2320)
[https://github.com/cockroachdb/cockroach/blob/2c1d4afd2c2de091a00a4dba4392ceeaf8504a95/pkg/sql/conn_executor_prepare.go#L477-L479](https://github.com/cockroachdb/cockroach/blob/2c1d4afd2c2de091a00a4dba4392ceeaf8504a95/pkg/sql/conn_executor_prepare.go#L477-L479)
[https://github.com/cockroachdb/cockroach/blob/2c1d4afd2c2de091a00a4dba4392ceeaf8504a95/pkg/sql/conn_executor_prepare.go#L468-L470](https://github.com/cockroachdb/cockroach/blob/2c1d4afd2c2de091a00a4dba4392ceeaf8504a95/pkg/sql/conn_executor_prepare.go#L468-L470)
[src/runtime/asm_arm64.s#L1171-L1173](src/runtime/asm_arm64.s#L1171-L1173)
[https://github.com/cockroachdb/cockroach/blob/2c1d4afd2c2de091a00a4dba4392ceeaf8504a95/pkg/sql/pgwire/conn.go#L309-L311](https://github.com/cockroachdb/cockroach/blob/2c1d4afd2c2de091a00a4dba4392ceeaf8504a95/pkg/sql/pgwire/conn.go#L309-L311)
[https://github.com/cockroachdb/cockroach/blob/2c1d4afd2c2de091a00a4dba4392ceeaf8504a95/pkg/sql/conn_executor.go#L899-L901](https://github.com/cockroachdb/cockroach/blob/2c1d4afd2c2de091a00a4dba4392ceeaf8504a95/pkg/sql/conn_executor.go#L899-L901)
[https://github.com/cockroachdb/cockroach/blob/2c1d4afd2c2de091a00a4dba4392ceeaf8504a95/pkg/sql/conn_executor.go#L2102-L2104](https://github.com/cockroachdb/cockroach/blob/2c1d4afd2c2de091a00a4dba4392ceeaf8504a95/pkg/sql/conn_executor.go#L2102-L2104)
[https://github.com/cockroachdb/cockroach/blob/2c1d4afd2c2de091a00a4dba4392ceeaf8504a95/pkg/sql/conn_executor.go#L2318-L2320](https://github.com/cockroachdb/cockroach/blob/2c1d4afd2c2de091a00a4dba4392ceeaf8504a95/pkg/sql/conn_executor.go#L2318-L2320)
[https://github.com/cockroachdb/cockroach/blob/2c1d4afd2c2de091a00a4dba4392ceeaf8504a95/pkg/sql/conn_executor_prepare.go#L477-L479](https://github.com/cockroachdb/cockroach/blob/2c1d4afd2c2de091a00a4dba4392ceeaf8504a95/pkg/sql/conn_executor_prepare.go#L477-L479)
[https://github.com/cockroachdb/cockroach/blob/2c1d4afd2c2de091a00a4dba4392ceeaf8504a95/pkg/sql/conn_executor_prepare.go#L460-L462](https://github.com/cockroachdb/cockroach/blob/2c1d4afd2c2de091a00a4dba4392ceeaf8504a95/pkg/sql/conn_executor_prepare.go#L460-L462)
[https://github.com/cockroachdb/cockroach/blob/2c1d4afd2c2de091a00a4dba4392ceeaf8504a95/pkg/sql/pgwire/pgwirebase/encoding.go#L841-L843](https://github.com/cockroachdb/cockroach/blob/2c1d4afd2c2de091a00a4dba4392ceeaf8504a95/pkg/sql/pgwire/pgwirebase/encoding.go#L841-L843)
</details>


```
src/runtime/asm_arm64.s in runtime.goexit at line 1172
pkg/sql/pgwire/conn.go in pkg/sql/pgwire.(*conn).processCommandsAsync.func1 at line 310
pkg/sql/conn_executor.go in pkg/sql.(*Server).ServeConn at line 900
pkg/sql/conn_executor.go in pkg/sql.(*connExecutor).run at line 2103
pkg/sql/conn_executor.go in pkg/sql.(*connExecutor).execCmd at line 2319
pkg/sql/conn_executor_prepare.go in pkg/sql.(*connExecutor).execBind at line 478
pkg/sql/conn_executor_prepare.go in pkg/sql.(*connExecutor).execBind.func2 at line 469
src/runtime/asm_arm64.s in runtime.goexit at line 1172
pkg/sql/pgwire/conn.go in pkg/sql/pgwire.(*conn).processCommandsAsync.func1 at line 310
pkg/sql/conn_executor.go in pkg/sql.(*Server).ServeConn at line 900
pkg/sql/conn_executor.go in pkg/sql.(*connExecutor).run at line 2103
pkg/sql/conn_executor.go in pkg/sql.(*connExecutor).execCmd at line 2319
pkg/sql/conn_executor_prepare.go in pkg/sql.(*connExecutor).execBind at line 478
pkg/sql/conn_executor_prepare.go in pkg/sql.(*connExecutor).execBind.func2 at line 461
pkg/sql/pgwire/pgwirebase/encoding.go in pkg/sql/pgwire/pgwirebase.DecodeDatum at line 842
```

### Tags

| Tag | Value |
| --- | --- |
| Command | start-single-node |
| Environment | v23.1.8 |
| Go Version | go1.19.10 |
| Platform | darwin arm64 |
| Distribution | CCL |
| Cockroach Release | v23.1.8 |
| Cockroach SHA | 2c1d4afd2c2de091a00a4dba4392ceeaf8504a95 |
| # of CPUs | 10 |
| # of Goroutines | 671 |



Jira issue: CRDB-33122

--- a/pkg/sql/pgwire/pgwirebase/encoding.go
+++ b/pkg/sql/pgwire/pgwirebase/encoding.go
@@ -313,9 +313,7 @@ func validateArrayDimensions(nDimensions int, nElements int) error {
 	return nil
 }
 
-// DecodeDatum decodes bytes with specified type and format code into
-// a datum. If res is nil, then user defined types are not attempted
-// to be resolved.
+// DecodeDatum decodes bytes with specified type and format code into a datum.
 func DecodeDatum(
 	ctx context.Context, evalCtx *eval.Context, typ *types.T, code FormatCode, b []byte,
 ) (tree.Datum, error) {
@@ -469,9 +467,10 @@ func DecodeDatum(
 			}
 			return &tree.DTSVector{TSVector: ret}, nil
 		}
-		if typ.Family() == types.ArrayFamily {
-			// Arrays come in in their string form, so we parse them as such and later
-			// convert them to their actual datum form.
+		switch typ.Family() {
+		case types.ArrayFamily, types.TupleFamily:
+			// Arrays and tuples come in in their string form, so we parse them
+			// as such and later convert them to their actual datum form.
 			if err := validateStringBytes(b); err != nil {
 				return nil, err
 			}

--- a/pkg/sql/pgwire/testdata/pgtest/tuple
+++ b/pkg/sql/pgwire/testdata/pgtest/tuple
@@ -127,6 +127,22 @@ ReadyForQuery
 {"Type":"CommandComplete","CommandTag":"SELECT 1"}
 {"Type":"ReadyForQuery","TxStatus":"I"}
 
+# Same as above, but using the text format (ParameterFormatCodes = [0] is for
+# text format).
+send
+Bind {"DestinationPortal": "p_valid_param_text", "PreparedStatement": "s_valid_param", "ParameterFormatCodes": [0], "Parameters": [{"text":"(t)"}], "ResultFormatCodes": [0]}
+Execute {"Portal": "p_valid_param_text"}
+Sync
+----
+
+until
+ReadyForQuery
+----
+{"Type":"BindComplete"}
+{"Type":"DataRow","Values":[{"text":"(t)"}]}
+{"Type":"CommandComplete","CommandTag":"SELECT 1"}
+{"Type":"ReadyForQuery","TxStatus":"I"}
+
 # empty parameter
 send
 Parse {"Name": "s_empty_param", "Query": "SELECT $1::r"}

